### PR TITLE
💪  Compute returns picked types!

### PR DESCRIPTION
**Pull request by nicolasgramlich**
Followup of: https://github.com/bluebirds-blue-jay/sequelize-service/pull/4

- [x] Code
- [x] Tests
- [x] Docs

### CODE REVIEW

**Comment by nicolasgramlich** [comment]
Rebasing...

**Comment by nicolasgramlich** [comment]
Reopening because github is a meanie...

File: src/classes/create-session.ts [hunk 1]
```diff
@@ -3,8 +3,8 @@ import { TCreateOptions } from '../types/create-options';
 import { ISequelizeService } from '../interfaces/sequelize-service';
 import { ICreateSession } from '../interfaces/create-session';
 
-export class CreateSession<W, R extends W, C> extends Session<W, R, C, TCreateOptions<R, C>> implements ICreateSession<W, R, C> {
-  public constructor(objects: W[], options: TCreateOptions<R, C>, service: ISequelizeService<W, R, C>) {
+export class CreateSession<W extends {}, R extends W, C extends {}, Compute extends keyof C> extends Session<W, R, C, TCreateOptions<R, C, Compute>> implements ICreateSession<W, R, C, Compute> {
```
Comment (SylvainEstevez) [inline src/classes/create-session.ts]: `Compute` should be shortened

File: src/classes/sequelize-service.ts [hunk 2]
```diff
@@ -67,15 +67,15 @@ export class SequelizeService<W, R extends W, C extends {} = {}> extends Service
...
         await this.executeHook(Hook.DID_CREATE, session, this._afterCreate.bind(this));
 
-        return session.getAt(0) as R & C;
+        return session.getAt(0) as R & Pick<C, Compute>;
       });
     });
   }
 
-  public async createMany(objects: W[], options: TCreateOptions<R, C> = {}): Promise<Collection<R & C>> {
+  public async createMany<Compute extends keyof C = undefined>(objects: W[], options: TCreateOptions<R, C, Compute> = {}): Promise<Collection<R & Pick<C, Compute>>> {
```
Comment (SylvainEstevez) [inline src/classes/sequelize-service.ts]: Feels a bit weird, doesn't it?

File: src/classes/sequelize-service.ts [hunk 3]
```diff
@@ -96,29 +96,29 @@ export class SequelizeService<W, R extends W, C extends {} = {}> extends Service
...
-  public async find(filters: TFilters<R>, options: TFindOptions<R, C> = {}): Promise<Collection<R & C>> {
+  public async find<Select extends keyof R, Compute extends keyof C = undefined>(filters: TFilters<R>, options: TFindOptions<R, C, Select, Compute> = {}): Promise<Collection<Pick<R, Select> & Pick<C, Compute>>> {
     const formattedFilters = this.toSequelizeWhere(filters);
     const sequelizeOptions = this.toSequelizeOptions<TSequelizeFindOptions<R>>(options, { where: formattedFilters });
 
     const objects = (await this.model.findAll(sequelizeOptions)).map(object => (<any>object).toJSON());
 
     if (objects.length) {
-      await this.computeProperties(new Session<W, R, C, TFindOptions<R, C>>(objects, options, this, filters));
+      await this.computeProperties(new Session<W, R, C, TFindOptions<R, C, Select, Compute>>(objects, options, this, filters));
```
Comment (SylvainEstevez) [inline src/classes/sequelize-service.ts]: Same here, generics should be 1-3 upper case letters

File: test/classes/sequelize-service.test.ts [hunk 4]
```diff
@@ -210,6 +209,38 @@ describe('SequelizeService', function () {
...
+      // user.date_of_birth; // Note: Should NOT compile!
+      // user.isAdult; // Note: Should NOT compile!
+    });
+
+    it('should not auto compute anything', async () => {
+      await userService.create({ email: 'foo', password: 'bar' });
+      const [ user ] = await userService.find({}, { select: [] });
+      expect(user).to.have.keys(['id']);
+      expect(user).to.not.have.keys(['isAdult']);
+      // user.age; // Note: Should NOT compile!
```
Comment (SylvainEstevez) [inline test/classes/sequelize-service.test.ts]: Type tests 💃 :)

File: src/classes/sequelize-service.ts (hunk 2 above)
Comment (nicolasgramlich) [reply src/classes/sequelize-service.ts]: Yeah, it looks super weird, but it does exactly what we want, lol

File: src/classes/create-session.ts (hunk 1 above)
Comment (nicolasgramlich) [reply src/classes/create-session.ts]: Ideas?

`RS` and `CS` ? 🤔 

Google's Java multi-letter generic convention is to put a `T` in the end, i.e. `ComputeT` and `SelectT`.

File: test/classes/sequelize-service.test.ts (hunk 4 above)
Comment (nicolasgramlich) [reply test/classes/sequelize-service.test.ts]: I wish, lol

File: src/classes/create-session.ts (hunk 1 above)
Comment (SylvainEstevez) [reply src/classes/create-session.ts]: RS and CS will do. As per having a better / more descriptive convention, I'm open to it but then all generics should follow it.

File: test/classes/sequelize-service.test.ts [hunk 5]
```diff
@@ -8,9 +8,8 @@ import { Collection } from '@bluejay/collection';
 import * as Sequelize from 'sequelize';
 import { Session } from '../../src/classes/session';
 import { SortOrder } from '../../src/constants/sort-order';
-import moment = require('moment');
-import { cat } from 'shelljs';
```
Comment (nicolasgramlich) [inline test/classes/sequelize-service.test.ts]: @SylvainEstevez btw, this was/is unused. Will remove.

**Comment by nicolasgramlich** [comment]
Rebased

File: src/classes/create-session.ts (hunk 1 above)
Comment (nicolasgramlich) [reply src/classes/create-session.ts]: Let's do `KR` and `KC` (read: `keys of R` and `keys of C`)

**Pull request merged**